src/Graph/GraphicalLasso.cpp

```diff
@@ -0,0 +1,194 @@
+/******************************************************************
+ * Created by Wei on 3/12/17.
+ *
+ * This is an implementation of this paper:
+ * "Sparse Inverse Covariance Estimation with the Graphical Lasso"
+ * 
+ * All names of parameters comply with the notaions in the paper.
+ ******************************************************************/
+
+#include "GraphicalLasso.hpp"
+
+#include <Eigen/Dense>
+#include <Eigen/SVD>
+#include <iostream>
+#include <map>
+
+#ifdef BAZEL
+#include "Models/LinearRegression.hpp"
+#include "Models/Model.hpp"
+#else
+#include "../Models/LinearRegression.hpp"
+#include "../Models/Model.hpp"
+#endif
+
+using namespace Eigen;
+using namespace std;
+
+GraphicalLasso::GraphicalLasso(const unordered_map<string, string>& opts) {
+    try {
+        tolerance = stod(opts.at("tolerance"));
+    } catch(std::out_of_range& oor) {
+        tolerance = default_tolerance;
+    }
+    try {
+        learningRate = stod(opts.at("learning_rate"));
+    } catch(std::out_of_range& oor) {
+        learningRate = default_learning_rate;
+    }
+}
+
+
+GraphicalLasso::GraphicalLasso() {
+    learningRate = default_learning_rate;
+    tolerance = default_tolerance;
+}
+
+
+/*void GraphicalLasso::stop() {
+    shouldStop = true;
+}*/
+
+void GraphicalLasso::setTolerance(float tol) {
+    tolerance = tol;
+}
+
+void GraphicalLasso::assertReadyToRun() {
+    return;    // there is no data that cannot be inferred
+}
+
+
+void GraphicalLasso::setUpRun() {
+    mtx.lock();
+    isRunning = true;
+    progress = 0.0;
+    shouldStop = false;
+}
+
+void GraphicalLasso::finishRun() {
+    isRunning = false;
+    progress = 1.0;
+    mtx.unlock();
+}
+
+void GraphicalLasso::setLearningRate(float lr) {
+    learningRate = lr;
+}
+
+MatrixXf GraphicalLasso::matrixRaiseToHalf(MatrixXf& matrix) {
+    JacobiSVD<MatrixXf> svd(matrix, ComputeThinU | ComputeThinV);
+    MatrixXf D = svd.singularValues().array().sqrt().matrix().asDiagonal();
+    MatrixXf U = svd.matrixU();
+    return U * D * U.transpose();
+}
+
+MatrixXf GraphicalLasso::pseudoInverse(MatrixXf& matrix) {
```

Comment:
Please put this into Math.cpp (In Mach folder) since this is a general math solution, not specific to Glasso.